### PR TITLE
Optimise QgsAbstractGeometry::snappedToGrid

### PR DESCRIPTION
**Pull request by nyalldawson**
- Optimise QgsLineString::snappedToGrid: The old approach was quite inefficient and resulted in many reallocations and array resizing
- Add option to remove redundant vertices in QgsAbstractGeometry::snappedToGrid. If opted in, then vertices which form a midpoint of a straight line segment will be removed in the output. This can result in substantially smaller geometries, as the current behavior will leave these redundant vertices in place.

This is leading toward on-the-fly simplication of labeling masks, for https://github.com/qgis/QGIS/issues/54788 and https://github.com/qgis/QGIS/issues/50543

### CODE REVIEW

File: src/core/geometry/qgslinestring.cpp [hunk 1]
```diff
@@ -326,17 +326,104 @@ bool QgsLineString::isValid( QString &error, Qgis::GeometryValidityFlags flags )
...
+      const bool isPointEqual = qgsDoubleNear( roundedX, previousX )
+                                && qgsDoubleNear( roundedY, previousY )
+                                && ( !hasZ || dSpacing <= 0 || qgsDoubleNear( roundedZ, previousZ ) )
+                                && ( !hasM || mSpacing <= 0 || qgsDoubleNear( roundedM, previousM ) );
+      if ( isPointEqual )
+        continue;
+
+      // maybe previous point is redundant and is just a midpoint on a straight line -- let's check
+      bool previousPointRedundant = false;
+      if ( outSize > 1 && !hasZ && !hasM )
```
Comment (uclaros) [inline src/core/geometry/qgslinestring.cpp:382]: Should we not be removing points that end up having the same `x,y` coordinates but different `z` or `m` as redundant?
I'd even argue that we should be removing those duplicates even with `removeReduntantPoints == false`!

File: src/core/geometry/qgslinestring.cpp [hunk 2]
```diff
@@ -326,17 +326,104 @@ bool QgsLineString::isValid( QString &error, Qgis::GeometryValidityFlags flags )
...
+    const double currentY = *yIn++;
+    const double currentZ = zIn ? *zIn++ : 0;
+    const double currentM = mIn ? *mIn++ : 0;
+
+    const double roundedX = hSpacing > 0 ? ( std::round( currentX / hSpacing ) * hSpacing ) : currentX;
+    const double roundedY = vSpacing > 0 ? ( std::round( currentY / vSpacing ) * vSpacing ) : currentY;
+    const double roundedZ = hasZ && dSpacing > 0 ? ( std::round( currentZ / dSpacing ) * dSpacing ) : currentZ;
+    const double roundedM = hasM && mSpacing > 0 ? ( std::round( currentM / mSpacing ) * mSpacing ) : currentM;
+
+    if ( i > 0 )
```
Comment (uclaros) [inline src/core/geometry/qgslinestring.cpp:371]: I think that placing the first iteration first:
```suggestion
    if ( i == 0 )
```
will make it easier to follow

**Review by uclaros — COMMENTED** [review]
I'm curious about the actual impact of this on a release build, do you have any before/after numbers?
Also, why not update `QgsCurve::snapToGridPrivate()` with this new implementation?

File: src/core/geometry/qgslinestring.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/core/geometry/qgslinestring.cpp:371]: Done

**Comment by nyalldawson** [comment]
@uclaros 
> I'm curious about the actual impact of this on a release build, do you have any before/after numbers?

For a moderately complex linestring, simplifying such that the number of vertices is reduced by roughly half, this PR cuts about 10% off the execution time. (The benefits are greater when there's less vertices removed, and basically zero when most of the vertices get removed).

> Also, why not update QgsCurve::snapToGridPrivate() with this new implementation?

Good point, I've done that now. I've disabled the removal of redundant points for circular strings though, as that logic doesn't apply for those.

File: src/core/geometry/qgslinestring.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/geometry/qgslinestring.cpp:382]: > Should we not be removing points that end up having the same x,y coordinates but different z or m as redundant?
I'd even argue that we should be removing those duplicates even with removeReduntantPoints == false!

I don't think so. At least, I'd argue that we should only do that if the z/m value of the point would be the same if it were linearly interpolated from the vertices left intact at the original redundant point x/y. And that logic would be non-trivial and isn't relevant for my pressing use case (https://github.com/qgis/QGIS/pull/57623).

**Comment by uclaros** [comment]
> For a moderately complex linestring, simplifying such that the number of vertices is reduced by roughly half, this PR cuts about 10% off the execution time.

But then the gain comes from the actual geometry simplification. I was referring to the first optimization bullet, ie. avoiding reallocating and resizing the coordinates' vectors.

**Comment by nyalldawson** [comment]
> But then the gain comes from the actual geometry simplification. I was referring to the first optimization bullet, ie. avoiding reallocating and resizing the coordinates' vectors.

No, that's without the redundant point removal for both tests. Even without redundant point removal then exactly duplicate points are still removed.

**Comment by nyalldawson** [comment]
@uclaros am i good to merge now?

**Review by uclaros — APPROVED** [review]

**Comment by nyalldawson** [comment]
Thanks @uclaros !

**Pull request merged**